README.md
# scheduleplus

A scheduler built on cron syntax with workday/holiday check feature

| Atributes  | Allowed values | Allowed special characters | Required |
| ---------- | -------------- | -------------------------- | -------- |
| Minutes    | 0-59           | \* , - /                   | Yes      |
| Hours      | 0-23           | \* , - /                   | Yes      |
| Days       | 1-31           | \* , - /                   | Yes      |
| Months     | 1-12, JAN-DEC  | \* , - /                   | Yes      |
| Weekdays   | 0-6, MON-SUN   | \* , - / F L               | Yes      |
| Holiday(\* | 0, 1           | \*                         | No       |

### Holiday\*

0 = checking if workday, 1 = checking if holiday, \* = any day

## Installation

```
pip install scheduleplus
```

## Usage

### Running a function on schedule

This runs the function every 5 minutes.

```python
from scheduleplus.schedule import Scheduler

def work(data):
    print(f"Working {data}...")

schedule = Scheduler()

schedule.cron("*/5 * * * *").do_function(work, "test")

while True:
    schedule.run_function_jobs()
    time.sleep(1)
```

### Running a callback

This runs the callback with a dictonary every 5 minutes.

```python
from scheduleplus.schedule import Scheduler

def work(data):
    print(f"Working {data}...")

schedule = Scheduler()

schedule.cron("*/5 * * * *").do_callback({"message": "cool"})

while True:
    schedule.run_callback_jobs(work)
    time.sleep(1)
```

### List jobs

```python
schedule = Scheduler()

schedule.cron("*/5 * * * *").do_function(work, "test")
schedule.cron("1 1 1 1 1").do_callback({"message": "cool"})

schedule.list_jobs()
```

Result

```
Job id   Cron          Next run time         Time left             Function        Callback message
-------- ------------- --------------------- --------------------- --------------- ---------------------
1        */5 * * * *   2022-11-04 13:50:00   0:00:29               work('test')
2        1 1 1 1 1     2030-01-01 01:01:00   2614 days, 11:11:29                   {'message': 'cool'}
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Changed

## [0.1.1] - 2022-11-08

- Added support for Python 3.10

## [0.1.0] - 2022-11-08

- First release


tests/test_parse_part.py
from scheduleplus.cronparser import CronParser


def _run_test(part_str, part):
    return CronParser("* * * * *")._parse_part(part_str, part)


def test_parse_minute():
    part = 0
    assert _run_test("*", part) == [item for item in range(0, 60)]
    assert _run_test("*/2", part) == [item for item in range(0, 60, 2)]
    assert _run_test("*/30", part) == [0, 30]
    assert _run_test("1-10/2", part) == [item for item in range(1, 11, 2)]
    assert _run_test("1/2", part) == [item for item in range(1, 60, 2)]
    assert _run_test("10-20/9", part) == [10, 19]
    assert _run_test("10-10/9", part) == [10]
    assert _run_test("10,59", part) == [10, 59]
    assert _run_test("1,2,3,4,5,10,11,12,13", part) == [1, 2, 3, 4, 5, 10, 11, 12, 13]
    assert _run_test("10-10,20-30/2", part) == [10, 20, 22, 24, 26, 28, 30]
    assert _run_test("55/2", part) == [55, 57, 59]
    assert _run_test("40,55", part) == [40, 55]
    assert _run_test("40,55,40,55", part) == [40, 55]
    assert _run_test("55,40", part) == [40, 55]
    assert _run_test("0-5/2,10-12,20,22", part) == [0, 2, 4, 10, 11, 12, 20, 22]
    assert _run_test("100", part) == []


def test_parse_hour():
    part = 1
    assert _run_test("*", part) == [item for item in range(0, 24)]
    assert _run_test("24", part) == []
    assert _run_test("23", part) == [23]


def test_parse_day():
    part = 2
    assert _run_test("*", part) == [item for item in range(1, 32)]
    assert _run_test("32", part) == []
    assert _run_test("31", part) == [31]


def test_parse_month():
    part = 3
    assert _run_test("*", part) == [item for item in range(1, 13)]
    assert _run_test("13", part) == []
    assert _run_test("12", part) == [12]
    assert _run_test("DeC", part) == [12]
    assert _run_test("JaN/1", part) == [item for item in range(1, 13)]


def test_parse_weekday():
    part = 4
    assert _run_test("*", part) == [item for item in range(0, 7)]
    assert _run_test("7", part) == []
    assert _run_test("6", part) == [6]
    assert _run_test("MoN-tUe", part) == [0, 1]
    assert _run_test("MoN-fRi/2", part) == [0, 2, 4]
    assert _run_test("MoN-fRi/10", part) == [0]


tests/test_text_table.py
from scheduleplus.text_table import table


def test_text_table():
    cols = ("c1", "c2")
    data = [("d1", "d_2", "d_2")]
    table(cols, data)


tests/test_parse_minute.py
from scheduleplus.util import parse_minute


def test_parse_minute():
    assert parse_minute("*") == [item for item in range(0, 60)]
    assert parse_minute("*/2") == [item for item in range(0, 60, 2)]
    assert parse_minute("1-10/2") == [item for item in range(1, 11, 2)]
    assert parse_minute("1/2") == [item for item in range(1, 60, 2)]
    assert parse_minute("10-20/9") == [10, 19]
    assert parse_minute("10-10/9") == [10]
    assert parse_minute("10,59") == [10, 59]
    assert parse_minute("1,2,3,4,5,10,11,12,13") == [1, 2, 3, 4, 5, 10, 11, 12, 13]
    assert parse_minute("10-10,20-30/2") == [10, 20, 22, 24, 26, 28, 30]
    assert parse_minute("55/2") == [55, 57, 59]
    assert parse_minute("40,55") == [40, 55]
    assert parse_minute("40,55,40,55") == [40, 55]
    assert parse_minute("55,40") == [40, 55]
    assert parse_minute("0-5/2,10-12,20,22") == [0, 2, 4, 10, 11, 12, 20, 22]


src/scheduleplus/util.py
import datetime
from datetime import timedelta


RANGES = [
    (0, 59),
    (0, 23),
    (1, 31),
    (1, 12),
    (0, 6),
]

ATTRIBUTES = [
    "minute",
    "hour",
    "day",
    "month",
    "isoweekday",
]

ALIASES = {
    "@yearly": "0 0 1 1 *",
    "@annually": "0 0 1 1 *",
    "@monthly": "0 0 1 * *",
    "@weekly": "0 0 * * 0",
    "@daily": "0 0 * * *",
    "@hourly": "0 * * * *",
}

ISODAYS = {"sun": 0, "mon": 1, "tue": 2, "wed": 3, "thu": 4, "fri": 5, "sat": 6}
MONTHS = {
    "jan": 1,
    "feb": 2,
    "mar": 3,
    "apr": 4,
    "may": 5,
    "jun": 6,
    "jul": 7,
    "aug": 8,
    "sep": 9,
    "oct": 10,
    "nov": 11,
    "dec": 12,
}


def _tris_extra_whitespace(str):
    return " ".join(str.split())


def _parse_cron_str(cron_str: str):
    parsed_cron_data = []
    cron_str = _tris_extra_whitespace(cron_str)
    cron_str = _replace_alias(cron_str)
    for part_index, part_str in enumerate(cron_str.split(" ")):
        parsed_cron_data.append(_parse_part(part_str, part_index))
    return parsed_cron_data


def _replace_alias(cron_str: str):
    cron_str = cron_str.lower()
    for index in ALIASES:
        if index in cron_str:
            cron_str = ALIASES[index]
    return cron_str


def _replace_month_day(part_str: str, part_index: int):
    part_str = part_str.lower()
    if part_index == 4:
        for index, day in enumerate(ISODAYS):
            part_str = part_str.replace(day, str(index))
    elif part_index == 3:
        for index, month in enumerate(MONTHS):
            part_str = part_str.replace(month, str(index + 1))
    return part_str


def _parse_part(part_str: str, part_index: int):
    part_str = _replace_month_day(part_str, part_index)
    return_list = []
    for part in part_str.split(","):
        start, end, step = RANGES[part_index][0], RANGES[part_index][1] + 1, 1
        if part == "*":
            for num in range(start, end, step):
                if num not in return_list:
                    return_list.append(num)
        elif "/" in part:
            tmp = part.split("/")
            if tmp[0] != "*":
                if "-" in tmp[0]:
                    tmp2 = tmp[0].split("-")
                    start, end = int(tmp2[0]), int(tmp2[1]) + 1
                else:
                    start, end = int(tmp[0]), RANGES[part_index][1] + 1
            step = int(tmp[1])
            for num in range(start, end, step):
                if num not in return_list:
                    return_list.append(num)
        elif "-" in part:
            tmp2 = part.split("-")
            start, end = int(tmp2[0]), int(tmp2[1]) + 1
            for num in range(start, end, step):
                if num not in return_list:
                    return_list.append(num)
        else:
            if int(part) not in return_list and int(part) in range(
                RANGES[part_index][0], RANGES[part_index][1] + 1
            ):
                return_list.append(int(part))
    return_list.sort()
    return return_list


def _proc_minute(index, parsed_data, next_run_time: datetime.datetime):
    for num in parsed_data[index]:
        if num >= next_run_time.minute:
            next_run_time = next_run_time.replace(minute=num)
            break
    else:
        next_run_time = next_run_time + timedelta(hours=1)
        next_run_time = next_run_time.replace(minute=0)
        next_run_time = _proc_minute(index, parsed_data, next_run_time)
    return next_run_time


def _proc_hour(index, parsed_data, next_run_time: datetime.datetime):
    for num in parsed_data[index]:
        if num >= next_run_time.hour:
            next_run_time = next_run_time.replace(hour=num)
            break
    else:
        next_run_time = next_run_time + timedelta(days=1)
        next_run_time = next_run_time.replace(minute=0)
        next_run_time = next_run_time.replace(hour=0)
        next_run_time = _proc_minute(0, parsed_data, next_run_time)
        next_run_time = _proc_hour(index, parsed_data, next_run_time)
    return next_run_time


def _proc_day(index, parsed_data, next_run_time: datetime.datetime):
    for num in parsed_data[index]:
        if num >= next_run_time.day:
            next_run_time = next_run_time.replace(day=num)
            break
    else:
        next_run_time = next_run_time.replace(day=1)
        next_run_time = next_run_time + timedelta(days=32)
        next_run_time = next_run_time.replace(day=1)
        next_run_time = next_run_time.replace(minute=0)
        next_run_time = next_run_time.replace(hour=0)
        next_run_time = _proc_minute(0, parsed_data, next_run_time)
        next_run_time = _proc_hour(1, parsed_data, next_run_time)
        next_run_time = _proc_day(index, parsed_data, next_run_time)
    return next_run_time


def _incr_year(next_run_time):
    mod = next_run_time.year % 4
    if mod == 0 and (next_run_time.month, next_run_time.day) < (2, 29):
        return timedelta(days=365 + 1)
    if mod == 3 and (next_run_time.month, next_run_time.day) > (2, 29):
        return timedelta(days=365 + 1)
    return timedelta(days=365)


def _proc_month(index, parsed_data, next_run_time: datetime.datetime):
    for num in parsed_data[index]:
        if num >= next_run_time.month:
            next_run_time = next_run_time.replace(month=num)
            break
    else:
        next_run_time = next_run_time.replace(month=1)
        next_run_time = next_run_time + _incr_year(next_run_time)
        next_run_time = next_run_time.replace(month=1)
        next_run_time = next_run_time.replace(day=1)
        next_run_time = next_run_time.replace(minute=0)
        next_run_time = next_run_time.replace(hour=0)
        next_run_time = _proc_minute(0, parsed_data, next_run_time)
        next_run_time = _proc_hour(1, parsed_data, next_run_time)
        next_run_time = _proc_day(2, parsed_data, next_run_time)
        next_run_time = _proc_month(index, parsed_data, next_run_time)
    return next_run_time


def _proc_weekday(index, parsed_data, next_run_time: datetime.datetime, cron_str):
    if next_run_time.weekday() in parsed_data[index]:
        return next_run_time
    for num in parsed_data[index]:
        num = num - 1
        if num < 0:
            num = 6
        if num == next_run_time.weekday():
            break
        else:
            while num != next_run_time.weekday():
                next_run_time = next_run_time + timedelta(days=1)
                next_run_time = next_run_time.replace(minute=0)
                next_run_time = next_run_time.replace(hour=0)
                next_run_time = _get_next_run_time(cron_str, now=next_run_time)
    return next_run_time


def _get_next_run_time(cron_str: str, now=None):
    parsed_data = _parse_cron_str(cron_str)
    if not now:
        now = datetime.datetime.now()
    next_run_time = now + timedelta(minutes=1)
    next_run_time = next_run_time.replace(second=0, microsecond=0)
    for index in range(0, 5):
        if index == 0:
            next_run_time = _proc_minute(index, parsed_data, next_run_time)
        if index == 1:
            next_run_time = _proc_hour(index, parsed_data, next_run_time)
        if index == 2:
            next_run_time = _proc_day(index, parsed_data, next_run_time)
        if index == 3:
            next_run_time = _proc_month(index, parsed_data, next_run_time)
        if index == 4:
            next_run_time = _proc_weekday(index, parsed_data, next_run_time, cron_str)
    return next_run_time


def _iter(cron_str: str):
    now = datetime.datetime.now()
    for _ in range(1, 11):
        now = _get_next_run_time(cron_str, now)
        yield now


if __name__ == "__main__":
    cron = _iter("* * * * *")
    for row in cron:
        print(row)


src/scheduleplus/schedule.py
import datetime
import json

from workalendar.europe import Sweden

from scheduleplus.cronparser import CronParser
from scheduleplus.text_table import table


class Job(object):
    _cron_str = ""
    _func = None
    _callback_message = None
    _id = None

    def __init__(self, job_id, cron_str, now=None, wk_country=None) -> None:
        if not wk_country:
            wk_country = Sweden
        self._id = job_id
        self._wk_country = wk_country
        self._cron_str = cron_str
        self._cron_parser = CronParser(self._cron_str, now=now, wk_country=wk_country)

    def do_function(self, func, *args, **kwargs):
        self._func = func
        self._args = args
        self._kwargs = kwargs

    def do_callback(self, message: dict):
        self._callback_message = message

    def run_job_meta(self, callback):
        if (
            self._callback_message
            and self._cron_parser.next_run_time() <= datetime.datetime.now()
        ):
            self._cron_parser.iter()
            callback(self._callback_message)

    def run_job_func(self):
        if self._func and self._cron_parser.next_run_time() <= datetime.datetime.now():
            self._cron_parser.iter()
            self._func(*self._args, **self._kwargs)

    def next_run(self):
        return self._cron_parser.next_run_time()

    def time_left(self):
        return str(self._cron_parser.next_run_time() - datetime.datetime.now()).split(
            "."
        )[0]

    def func_name(self):
        if self._func:
            func_name = self._func.__name__
            args = str(self._args).replace("(", "").replace(")", "")
            kwargs = (
                str(self._kwargs)
                .replace("{", "")
                .replace("}", "")
                .replace(": ", "=")
                .replace("'", "")
            )
            if kwargs:
                args += ", "
            return func_name + "(" + args + kwargs + ")"
        return ""

    def cron_str(self):
        return self._cron_str

    def callback_message(self):
        return self._callback_message if self._callback_message else ""

    def id(self):
        return self._id


class Scheduler(object):
    _jobs = []
    _next_job_id = 1

    def __init__(self):
        pass

    def cron(self, cron_str, now=None, wk_country=None):
        if not wk_country:
            wk_country = Sweden
        self._wk_country = wk_country
        job_id = self._next_job_id
        self._next_job_id += 1
        job = Job(job_id, cron_str, now, wk_country)
        self._jobs.append(job)
        self._jobs.sort(key=lambda a: a.next_run())
        return job

    def run_function_jobs(self):
        for job in self._jobs:
            job.run_job_func()

    def run_callback_jobs(self, callback):
        for job in self._jobs:
            job.run_job_meta(callback)

    def dict(self):
        data = {
            "columns": [
                "Job id",
                "Cron",
                "Next run time",
                "Time left",
                "Function",
                "Callback message",
            ],
            "data": [],
        }
        for job in self._jobs:
            data["data"].append(
                (
                    str(job.id()),
                    str(job.cron_str()),
                    str(job.next_run()),
                    str(job.time_left()),
                    str(job.func_name()),
                    str(job.callback_message()),
                )
            )
        return data

    def json(self):
        return json.dumps(self.dict())

    def list_jobs(self):
        data = self.dict()
        print(table(data["columns"], data["data"]))


src/scheduleplus/text_table.py
EXTRA_SPACE = 3


def table_line(width):
    return "".zfill(width - 1).replace("0", "-") + " "


def table(column_names, table_data):
    col_sizes = {}
    for index, col in enumerate(column_names):
        if index not in col_sizes:
            col_sizes[index] = len(col) + EXTRA_SPACE
        if len(col) + EXTRA_SPACE > col_sizes[index]:
            col_sizes[index] = len(col) + EXTRA_SPACE

    for row in table_data:
        for index, col in enumerate(row):
            if index not in col_sizes:
                col_sizes[index] = len(col) + EXTRA_SPACE
            if len(col) + EXTRA_SPACE > col_sizes[index]:
                col_sizes[index] = len(col) + EXTRA_SPACE

    header = ""
    for index, col in enumerate(column_names):
        header += f"{col:{col_sizes[index]}}"
    header += "\n"
    for col_size in col_sizes:
        header += table_line(col_sizes[col_size])
    header += "\n"

    data = ""
    for row in table_data:
        for index, col in enumerate(row):
            data += f"{col:{col_sizes[index]}}"
        data += "\n"

    return header + data


src/scheduleplus/cronparser.py
import datetime
from datetime import timedelta

from workalendar.europe import Sweden

RANGES = [
    (0, 59),
    (0, 23),
    (1, 31),
    (1, 12),
    (0, 6),
    (0, 1),
]

ATTRIBUTES = [
    "minute",
    "hour",
    "day",
    "month",
    "weekday",
]

ALIASES = {
    "@yearly": "0 0 1 1 *",
    "@annually": "0 0 1 1 *",
    "@monthly": "0 0 1 * *",
    "@weekly": "0 0 * * 0",
    "@daily": "0 0 * * *",
    "@hourly": "0 * * * *",
}

ISODAYS = {
    "mon": 0,
    "tue": 1,
    "wed": 2,
    "thu": 3,
    "fri": 4,
    "sat": 5,
    "sun": 6,
}

MONTHS = {
    "jan": 1,
    "feb": 2,
    "mar": 3,
    "apr": 4,
    "may": 5,
    "jun": 6,
    "jul": 7,
    "aug": 8,
    "sep": 9,
    "oct": 10,
    "nov": 11,
    "dec": 12,
}


class CronParser:
    def __init__(self, cron_str: str, now=None, wk_country=None):
        if not wk_country:
            wk_country = Sweden()
        self._wk_country = wk_country
        # self._wk_country = self._wk_country
        self._cron_str = cron_str
        self._next_run_time = datetime.datetime.now()
        if now:
            self._next_run_time = now
        self._next_run_time = self._next_run_time.replace(second=0, microsecond=0)
        self._next_run_time += timedelta(minutes=1)

    def _trim_extra_whitespace(self):
        return " ".join(self._cron_str.split())

    def _parse_cron_str(self):
        parsed_cron_data = []
        self._cron_str = self._trim_extra_whitespace()
        self._cron_str = self._replace_alias()
        for part_index, part_str in enumerate(self._cron_str.split(" ")):
            parsed_cron_data.append(self._parse_part(part_str, part_index))
        return parsed_cron_data

    def _replace_alias(self):
        cron_str = self._cron_str.lower()
        for index in ALIASES:
            if index in cron_str:
                cron_str = ALIASES[index]
        return cron_str

    def _replace_month_day(self, part_str: str, part_index: int):
        part_str = part_str.lower()
        if part_index == 4:
            for index, day in enumerate(ISODAYS):
                part_str = part_str.replace(day, str(index))
        elif part_index == 3:
            for index, month in enumerate(MONTHS):
                part_str = part_str.replace(month, str(index + 1))
        return part_str

    def _parse_part(self, part_str: str, part_index: int):
        part_str = self._replace_month_day(part_str, part_index)
        return_list = []
        for part in part_str.split(","):
            start, end, step = RANGES[part_index][0], RANGES[part_index][1] + 1, 1
            if part == "*":
                for num in range(start, end, step):
                    if num not in return_list:
                        return_list.append(num)
            elif "/" in part:
                tmp = part.split("/")
                if tmp[0] != "*":
                    if "-" in tmp[0]:
                        tmp2 = tmp[0].split("-")
                        start, end = int(tmp2[0]), int(tmp2[1]) + 1
                    else:
                        start, end = int(tmp[0]), RANGES[part_index][1] + 1
                step = int(tmp[1])
                for num in range(start, end, step):
                    if num not in return_list:
                        return_list.append(num)
            elif "-" in part:
                tmp2 = part.split("-")
                start, end = int(tmp2[0]), int(tmp2[1]) + 1
                for num in range(start, end, step):
                    if num not in return_list:
                        return_list.append(num)
            elif part_str.upper() == "F":
                return_list.append("F")
            elif part_str.upper() == "L":
                return_list.append("L")
            else:
                if int(part) not in return_list and int(part) in range(
                    RANGES[part_index][0], RANGES[part_index][1] + 1
                ):
                    return_list.append(int(part))
        return_list.sort()
        return return_list

    def _proc_minute(self, index, parsed_data):
        for num in parsed_data[index]:
            if num >= self._next_run_time.minute:
                self._next_run_time = self._next_run_time.replace(minute=num)
                break
        else:
            self._next_run_time = self._next_run_time + timedelta(hours=1)
            self._next_run_time = self._next_run_time.replace(minute=0)
            self._next_run_time = self._proc_minute(index, parsed_data)
        return self._next_run_time

    def _proc_hour(self, index, parsed_data):
        for num in parsed_data[index]:
            if num >= self._next_run_time.hour:
                self._next_run_time = self._next_run_time.replace(hour=num)
                break
        else:
            self._next_run_time = self._next_run_time + timedelta(days=1)
            self._next_run_time = self._next_run_time.replace(minute=0)
            self._next_run_time = self._next_run_time.replace(hour=0)
            self._next_run_time = self._proc_minute(0, parsed_data)
            self._next_run_time = self._proc_hour(index, parsed_data)
        return self._next_run_time

    def _proc_day(self, index, parsed_data):
        if parsed_data[index] == ["L"]:
            self._next_run_time = self._next_run_time + timedelta(days=32)
            self._next_run_time = self._next_run_time.replace(day=1)
            self._next_run_time = self._next_run_time - timedelta(days=1)
        else:
            for num in parsed_data[index]:
                if num >= self._next_run_time.day:
                    self._next_run_time = self._next_run_time.replace(day=num)
                    break
            else:
                self._next_run_time = self._next_run_time.replace(day=1)
                self._next_run_time = self._next_run_time + timedelta(days=32)
                self._next_run_time = self._next_run_time.replace(day=1)
                self._next_run_time = self._next_run_time.replace(minute=0)
                self._next_run_time = self._next_run_time.replace(hour=0)
                self._next_run_time = self._proc_minute(0, parsed_data)
                self._next_run_time = self._proc_hour(1, parsed_data)
                self._next_run_time = self._proc_day(index, parsed_data)
        return self._next_run_time

    def _incr_year(self):
        mod = self._next_run_time.year % 4
        if mod == 0 and (self._next_run_time.month, self._next_run_time.day) < (2, 29):
            return timedelta(days=365 + 1)
        if mod == 3 and (self._next_run_time.month, self._next_run_time.day) > (2, 29):
            return timedelta(days=365 + 1)
        return timedelta(days=365)

    def _proc_month(self, index, parsed_data):
        for num in parsed_data[index]:
            if num >= self._next_run_time.month:
                self._next_run_time = self._next_run_time.replace(month=num)
                break
        else:
            self._next_run_time = self._next_run_time + self._incr_year()
            self._next_run_time = self._next_run_time.replace(month=1)
            self._next_run_time = self._next_run_time.replace(day=1)
            self._next_run_time = self._next_run_time.replace(minute=0)
            self._next_run_time = self._next_run_time.replace(hour=0)
            self._next_run_time = self._proc_minute(0, parsed_data)
            self._next_run_time = self._proc_hour(1, parsed_data)
            self._next_run_time = self._proc_day(2, parsed_data)
            self._next_run_time = self._proc_month(index, parsed_data)
        return self._next_run_time

    def _proc_weekday(self, index, parsed_data):
        if self._next_run_time.weekday() in parsed_data[index]:
            return self._next_run_time
        else:
            if parsed_data[index] == ["F"]:
                self._next_run_time = self._next_run_time.replace(day=1)
                while not self._wk_country.is_working_day(self._next_run_time.date()):
                    self._next_run_time = self._next_run_time + timedelta(days=1)
            elif parsed_data[index] == ["L"]:
                self._next_run_time = self._next_run_time + timedelta(days=32)
                self._next_run_time = self._next_run_time.replace(day=1)
                self._next_run_time = self._next_run_time - timedelta(days=1)
                while not self._wk_country.is_working_day(self._next_run_time.date()):
                    self._next_run_time = self._next_run_time - timedelta(days=1)
            else:
                while self._next_run_time.weekday() not in parsed_data[index]:
                    self._next_run_time = self._next_run_time + timedelta(days=1)
                    self._next_run_time = self._next_run_time.replace(minute=0)
                    self._next_run_time = self._next_run_time.replace(hour=0)
                    self._next_run_time = self._get_next_run_time()
        return self._next_run_time

    def _proc_holiday(self, index, parsed_data):
        if [0, 1] == parsed_data[index]:
            return self._next_run_time
        if 0 in parsed_data[index]:
            if self._wk_country.is_working_day(self._next_run_time.date()):
                return self._next_run_time
            else:
                while not self._wk_country.is_working_day(self._next_run_time.date()):
                    self._next_run_time = self._next_run_time + timedelta(days=1)
                    self._next_run_time = self._next_run_time.replace(minute=0)
                    self._next_run_time = self._next_run_time.replace(hour=0)
                    self._next_run_time = self._get_next_run_time()
        if 1 in parsed_data[index]:
            if self._wk_country.is_holiday(self._next_run_time.date()):
                return self._next_run_time
            else:
                while not self._wk_country.is_holiday(self._next_run_time.date()):
                    self._next_run_time = self._next_run_time + timedelta(days=1)
                    self._next_run_time = self._next_run_time.replace(minute=0)
                    self._next_run_time = self._next_run_time.replace(hour=0)
                    self._next_run_time = self._get_next_run_time()
        return self._next_run_time

    def _get_next_run_time(self):
        parsed_data = self._parse_cron_str()
        if len(parsed_data) < 5:
            raise ValueError("Error in cron string, not 5 or 6 values")
        for index in range(0, len(parsed_data)):
            if index == 0:
                self._next_run_time = self._proc_minute(index, parsed_data)
            if index == 1:
                self._next_run_time = self._proc_hour(index, parsed_data)
            if index == 2:
                self._next_run_time = self._proc_day(index, parsed_data)
            if index == 3:
                self._next_run_time = self._proc_month(index, parsed_data)
            if index == 4:
                self._next_run_time = self._proc_weekday(index, parsed_data)
            if index == 5:
                self._next_run_time = self._proc_holiday(index, parsed_data)
        return self._next_run_time

    def iter(self):
        self._next_run_time += timedelta(minutes=1)
        return self._get_next_run_time()

    def next_run_time(self):
        return self._get_next_run_time()
